src/main/java/com/sivalabs/bookstore/catalog/domain/ProductService.java

```diff
@@ -20,14 +21,37 @@ public ProductService(ProductRepository productRepository, ProductMapper product
     }
 
     public PagedResult<ProductModel> getProducts(int pageNo) {
-        int page = pageNo <= 1 ? 0 : pageNo - 1;
+        int page = getRepositoryPageNo(pageNo);
         Pageable pageable = PageRequest.of(page, PAGE_SIZE, Sort.Direction.ASC, "name");
         Page<ProductModel> productsPage =
                 productRepository.findAll(pageable).map(productMapper::toModel);
         return new PagedResult<>(productsPage);
     }
 
+    private int getRepositoryPageNo(int pageNo) {
+        int repositoryPage = pageNo <= 1 ? 0 : pageNo - 1;
+        return repositoryPage;
+    }
+
     public Optional<ProductModel> getProductByCode(String code) {
         return productRepository.findByCode(code).map(productMapper::toModel);
     }
+
+    public PagedResult<ProductModel> searchProductsByCriteria(String searchCriteria, int page) {
+        Page<ProductModel> productPage =
+                productRepository
+                        .findAllBy(
+                                textCriteriaMatching(searchCriteria),
+                                pageableOf(getRepositoryPageNo(page)))
+                        .map(productMapper::toModel);
+        return PagedResult.fromPage(productPage);
+    }
+
+    private TextCriteria textCriteriaMatching(String searchCriteria) {
+        return new TextCriteria().matchingPhrase(searchCriteria);
+    }
+
+    private Pageable pageableOf(int page) {
+        return PageRequest.of(page, PAGE_SIZE);
```

Comment:
While implementing pagination we should always specify a sort criteria, otherwise the results will be unpredictable due to lack of sort order.

Reply:
I know currently we are not adding weight to `name` vs `description` text search. In case we do, then sorting search results can change the order and make first search results less relevant. 

Having said that, I am not very familiar with Mongo TextSearch, if the above is not relevant, then I can add the sort criteria.

Reply:
There is no weight as of now, so `name` and `description` has equal priority while filtering. But we can apply sorting based on `name` to make the pagination results to have predictability.